src/processor.cc

```diff
@@ -16,92 +19,99 @@ StreamProcessor::StreamProcessor(size_t max_size_, bool doZS_, ProcessorType pro
 }  
 
 
+std::vector<uint32_t> bit_check(uint32_t word, uint32_t offset)
+{
+	std::vector<uint32_t> vect;
+	for (uint32_t i = 0; i<32; i++)
+	{
+		if (word & 1){
+			vect.push_back(i + offset);
+		}
+		word >>= 1;
+	}
+
+	return vect;
+}
+
 StreamProcessor::~StreamProcessor(){
 	//  fprintf(stderr,"Wrote %d muons \n",totcount);
 	myfile.close();
 }
 
-Slice* StreamProcessor::process(Slice& input, Slice& out)
-{
-	nbPackets++;
-	char* p = input.begin();
-	char* q = out.begin();
-	uint32_t counts = 0;
-	
-	if (processorType == ProcessorType::PASS_THROUGH) {
-		memcpy(q,p,input.size());
-		out.set_end(out.begin() + input.size());
-		out.set_counts(1);
-		return &out;
-	}
-	
+bool StreamProcessor::CheckFrameMultBlock(uint32_t inputSize){
+
 	int bsize = sizeof(block1);
-	if((input.size()-constants::orbit_trailer_size)%bsize!=0){
+	if((inputSize-20*constants::orbit_trailer_size - 32*20 -32)%bsize!=0){
 		LOG(WARNING)
-			<< "Frame size not a multiple of block size. Will be skipped. Size="
-			<< input.size() << " - block size=" << bsize;
-		return &out;
+			<< "Frame size not a multiple of block size. Will be skipped. Frame size = "
+			<< inputSize << ", block size = " << bsize;
+		return false;
 	}
+	return true;
+}
 
+std::vector<unsigned int> StreamProcessor::CountBX(Slice& input){
+
+	char* p = input.begin() + 32; // +32 to account for orbit header
+	std::vector<unsigned int> bx_vect;
+	while( p != input.end()){
+		block1 *bl = (block1*)(p);
+		if(bl->orbit[0]==constants::beefdead){ // found orbit trailer
+			orbit_trailer *ot = (orbit_trailer*)(p);
+			for (unsigned int k = 0; k < (14*8); k++){ // 14*8 = 14 frames, 8 links of orbit trailer containing BX hitmap
+				//bxcount += __builtin_popcount(ot->bx_map[k]);
+				std::vector<unsigned int> bx_vect_t = bit_check(ot->bx_map[k], k*32);
+				bx_vect.insert(bx_vect.end(), bx_vect_t.begin(), bx_vect_t.end());
+			}
+			return bx_vect;
+		}
+
+		p+=sizeof(block1);
+	}
+
+}
 
-	while(p!=input.end()){
-		bool brill_word = false;
-		bool endoforbit = false;
-		block1 *bl = (block1*)p;
+uint32_t StreamProcessor::FillOrbit(Slice& input, Slice& out, std::vector<unsigned int>* bx_vect){
```

Comment:
BXCount does a first pass of the orbit to find the trailer and then fill the vector with the correct BX numbers
Then FillOrbit actually does the filling of the memory using the correct BX numbers from the BX vector.
i.e two passes are required (because we need to use the info from the trailer before we can fill the output memory)
If the question is why is this a separate function rather than within the process function, this is because the process function was becoming extremely long and very nested, so I decided to break it up into a few functions. 
I didn't see any throughput issues in the lab test.

Reply:
I was missing the high-level explanation. Thanks for giving the explanation in the code.

About a separate function - of course it should be broken into smaller function(s), no objection here.

Reply:
Hi @tj710,

Would it be helpful if we already used the format of the packet header for the orbit header? My thinking is that we currently send the link header on all 8 links, but we could also just overwrite this with the intended packet header that contains the number of frames of BXs (even if we then have a proper packet header in addition). In that way you don't have to search for the trailer, but just jump there..

Reply:
Yes, however the firmware currently deployed at p5 does not have the packet header, so I went with an approach that works without it. This can be updated once the new firmware is deployed.